File: frontend/lib/api/business.ts
// lib/api/business.ts
// Maps to backend app/api/routes/business.py (prefix /business).
import { api } from './client';
import type { IdentityDraft, OperationsDraft } from '../../features/onboarding/Onboarding';

export interface DocumentUploadOut {
  id: number;
  doc_type: string;
  declared_owner_name: string;
  declared_business_name: string;
  declared_id_number: string;
  declared_expiry_date: string | null;
  validation_status: string;
  validation_notes: string;
  uploaded_at: string;
}

export interface VerificationStatusOut {
  verification_status: string;
  is_verified: boolean;
  tier: number;
  submitted_at: string | null;
  verification_date: string | null;
  recheck_date: string | null;
  onboarding_completed: boolean;
  documents: DocumentUploadOut[];
  has_submitted: boolean;
  missing_documents: string[];
  registered_name: string | null;
  display_name: string | null;
  business_type: string | null;
  industry_category: string | null;
  city: string | null;
  province: string | null;
  contact_person: string | null;
  contact_mobile: string | null;
  capabilities: string[];
  service_areas: string[];
  signup_intent: string;
  business_description: string | null;
}

export interface DashboardStatsOut {
  requirements_posted_count: number;
  quotations_submitted_count: number;
  requirements_awarded_count: number;
  member_since_year: number;
  profile_completion_pct: number;
  tier_hint: string;
}

export interface PublicBusinessProfileOut {
  id: number;
  registered_name: string | null;
  display_name: string | null;
  business_type: string | null;
  industry_category: string | null;
  city: string | null;
  province: string | null;
  business_description: string | null;
  capabilities: string[];
  service_areas: string[];
  is_verified: boolean;
  tier: number;
  verification_date: string | null;
  member_since_year: number;
  requirements_posted_count: number;
  requirements_awarded_count: number;
}

/** POST /business/onboarding — identity + operations, combined into the
 *  shape app/schemas/business.py::OnboardingSubmit expects. */
export async function submitOnboarding(identity: IdentityDraft, operations: OperationsDraft): Promise<void> {
  await api.post<void>('/business/onboarding', {
    registered_name: identity.registeredName,
    display_name: identity.displayName.trim() || null,
    business_type: identity.businessType,
    industry_category: identity.category,
    city: identity.city,
    province: identity.province,
    contact_person: identity.contactPerson,
    contact_mobile: identity.contactMobile,
    capabilities: operations.capabilities,
    service_areas: operations.serviceAreas,
    signup_intent: identity.signupIntent,
    business_description: operations.businessDescription.trim() || null,
  });
}

export interface DescriptionSuggestionOut {
  description: string | null;
}

/** POST /business/suggest-description — AI profile assistant's draft bio.
 *  A suggestion only: the caller decides whether to use it, same contract
 *  as extractDocumentFields and requirementsApi.suggestCategory. */
export async function suggestDescription(
  identity: IdentityDraft, capabilities: string[], serviceAreas: string[],
): Promise<DescriptionSuggestionOut> {
  return api.post<DescriptionSuggestionOut>('/business/suggest-description', {
    business_type: identity.businessType,
    industry_category: identity.category,
    capabilities,
    service_areas: serviceAreas,
    city: identity.city,
    province: identity.province,
  });
}

export interface UploadDocumentInput {
  docType: 'DTI' | 'SEC' | 'BIR' | 'MAYORS_PERMIT' | string;
  declaredOwnerName: string;
  declaredBusinessName: string;
  declaredIdNumber: string;
  declaredExpiryDate?: string; // YYYY-MM-DD
  file: File | Blob;
  fileName: string;
}

/** POST /business/documents — multipart/form-data. */
export async function uploadDocument(input: UploadDocumentInput): Promise<DocumentUploadOut> {
  const form = new FormData();
  form.append('doc_type', input.docType);
  form.append('declared_owner_name', input.declaredOwnerName);
  form.append('declared_business_name', input.declaredBusinessName);
  form.append('declared_id_number', input.declaredIdNumber);
  if (input.declaredExpiryDate) form.append('declared_expiry_date', input.declaredExpiryDate);
  form.append('file', input.file, input.fileName);
  return api.postForm<DocumentUploadOut>('/business/documents', form);
}

export interface DocumentExtractionOut {
  id_number: string | null;
}

/** POST /business/documents/extract — Assistive Document Extraction. A
 *  suggestion only: the caller pre-fills an editable field with the result,
 *  never submits it directly. `id_number: null` just means nothing was
 *  confidently found (no Gemini key configured, unreadable image, etc.) —
 *  never surfaced as an error, the field is just left for manual entry. */
export async function extractDocumentFields(docType: string, file: File | Blob, fileName: string): Promise<DocumentExtractionOut> {
  const form = new FormData();
  form.append('doc_type', docType);
  form.append('file', file, fileName);
  return api.postForm<DocumentExtractionOut>('/business/documents/extract', form);
}

/** POST /business/submit-for-verification */
export async function submitForVerification(): Promise<VerificationStatusOut> {
  return api.post<VerificationStatusOut>('/business/submit-for-verification');
}

/** GET /business/verification-status */
export async function getVerificationStatus(): Promise<VerificationStatusOut> {
  return api.get<VerificationStatusOut>('/business/verification-status');
}

/** GET /business/dashboard-stats */
export async function getDashboardStats(): Promise<DashboardStatsOut> {
  return api.get<DashboardStatsOut>('/business/dashboard-stats');
}

/** GET /business/{id}/profile — any other business's public-facing profile,
 *  e.g. "View buyer profile" from a requirement, or a respondent's card once
 *  quotations release. Never carries contact info — see PublicBusinessProfileOut. */
export async function getPublicProfile(businessId: number): Promise<PublicBusinessProfileOut> {
  return api.get<PublicBusinessProfileOut>(`/business/${businessId}/profile`);
}

File: backend/app/schemas/business.py
from datetime import date, datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, Field


class OnboardingSubmit(BaseModel):
    registered_name: str = Field(min_length=2, max_length=200)
    business_type: str = Field(min_length=2, max_length=100)
    industry_category: str = Field(min_length=2, max_length=100)
    city: str = Field(min_length=2, max_length=100)
    province: str = Field(min_length=2, max_length=100)
    contact_person: str = Field(min_length=2, max_length=120)
    contact_mobile: str = Field(min_length=7, max_length=20)
    capabilities: List[str] = Field(min_length=3, max_length=8)
    service_areas: List[str] = Field(min_length=1)
    signup_intent: Literal["FIND_SUPPLIERS", "FIND_WORK", "BOTH"] = "BOTH"


class DocumentUploadOut(BaseModel):
    id: int
    doc_type: str
    declared_owner_name: str
    declared_business_name: str
    declared_id_number: str
    declared_expiry_date: Optional[date]
    validation_status: str
    validation_notes: str
    uploaded_at: datetime


class VerificationStatusOut(BaseModel):
    verification_status: str
    is_verified: bool
    tier: int
    submitted_at: Optional[datetime] = None
    verification_date: Optional[datetime]
    recheck_date: Optional[datetime]
    onboarding_completed: bool
    documents: List[DocumentUploadOut]

    # True once the business has formally submitted — the point after which
    # they're waiting on us rather than us waiting on them.
    has_submitted: bool = False
    # Document types still needed before they can submit.
    missing_documents: List[str] = []

    # Echoed back so a returning business can resume onboarding without
    # relying on anything cached in the browser — business_type in
    # particular decides whether DTI or SEC registration is asked for.
    registered_name: Optional[str] = None
    business_type: Optional[str] = None
    industry_category: Optional[str] = None
    city: Optional[str] = None
    province: Optional[str] = None
    contact_person: Optional[str] = None
    contact_mobile: Optional[str] = None
    capabilities: List[str] = []
    service_areas: List[str] = []
    signup_intent: str = "BOTH"


class PublicBusinessProfileOut(BaseModel):
    """A business's public-facing profile — what any other logged-in business
    can see about them (e.g. "View buyer profile" from a requirement, or a
    respondent's card once quotations release). Deliberately excludes the
    login mobile number, contact person/mobile, and everything document- or
    admin-facing — those stay private to the business itself and to admin."""
    id: int
    registered_name: Optional[str]
    business_type: Optional[str]
    industry_category: Optional[str]
    city: Optional[str]
    province: Optional[str]
    capabilities: List[str] = []
    service_areas: List[str] = []
    is_verified: bool
    tier: int
    verification_date: Optional[datetime]
    member_since_year: int
    requirements_posted_count: int
    requirements_awarded_count: int


class AdminReviewAction(BaseModel):
    approve: bool
    notes: Optional[str] = Field(default=None, max_length=500)


class AdminFlaggedBusinessOut(BaseModel):
    user_id: int
    business_name: str
    mobile_number: str
    registered_name: Optional[str]
    business_type: Optional[str]
    industry_category: Optional[str]
    city: Optional[str]
    province: Optional[str]
    contact_person: Optional[str]
    contact_mobile: Optional[str]
    submitted_at: Optional[datetime]
    documents: List[DocumentUploadOut]


class AdminBusinessOut(BaseModel):
    """One row of the admin's full business roster — every registered
    business, not just the ones currently waiting on review."""
    user_id: int
    business_name: str
    mobile_number: str
    registered_name: Optional[str]
    business_type: Optional[str]
    industry_category: Optional[str]
    city: Optional[str]
    province: Optional[str]
    onboarding_completed: bool
    verification_status: str
    is_verified: bool
    tier: int
    submitted_at: Optional[datetime]
    created_at: datetime
    documents: List[DocumentUploadOut] = []


class DashboardStatsOut(BaseModel):
    requirements_posted_count: int
    quotations_submitted_count: int
    requirements_awarded_count: int  # times this business won as a supplier
    member_since_year: int
    profile_completion_pct: int
    tier_hint: str


class RegistrationsByDay(BaseModel):
    date: str  # YYYY-MM-DD
    count: int


class AdminStatsOut(BaseModel):
    total_businesses: int
    verified_businesses: int
    pending_review_count: int
    total_requirements: int
    open_requirements: int
    total_quotations: int
    registrations_last_7_days: List[RegistrationsByDay]